codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/requestcompression/RequestCompression.java

```diff
@@ -0,0 +1,166 @@
+
+    private static final String REQUEST_MIN_COMPRESSION_SIZE_BYTES = "RequestMinCompressSizeBytes";
+
+    private final List<RuntimeClientPlugin> runtimeClientPlugins = new ArrayList<>();
+
+    private static String getAddRequestCompressionMiddlewareFuncName(String operationName) {
+        return String.format("addOperation%sRequestCompressionMiddleware", operationName);
+    }
+
+    // Write operation plugin for request compression middleware
+    @Override
+    public void processFinalizedModel(GoSettings settings, Model model) {
+        ServiceShape service = settings.getService(model);
+        TopDownIndex.of(model)
+                .getContainedOperations(service).forEach(operation -> {
+                    if (!operation.hasTrait(RequestCompressionTrait.class)) {
+                        return;
+                    }
+                    SymbolProvider symbolProvider = GoCodegenPlugin.createSymbolProvider(model, settings);
+                    String funcName = getAddRequestCompressionMiddlewareFuncName(
+                            symbolProvider.toSymbol(operation).getName()
+                    );
+                    runtimeClientPlugins.add(RuntimeClientPlugin.builder().operationPredicate((m, s, o) -> {
+                        if (!o.hasTrait(RequestCompressionTrait.class)) {
+                            return false;
+                        }
+                        return o.equals(operation);
+                    }).registerMiddleware(MiddlewareRegistrar.builder()
+                        .resolvedFunction(SymbolUtils.createValueSymbolBuilder(funcName).build())
+                        .useClientOptions().build())
+                        .build());
+                });
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ServiceShape service = settings.getService(model);
+        for (ShapeId operationID : service.getAllOperations()) {
+            OperationShape operation = model.expectShape(operationID, OperationShape.class);
+            if (!operation.hasTrait(RequestCompressionTrait.class)) {
+                continue;
+            }
+            goDelegator.useShapeWriter(operation, writeMiddlewareHelper(symbolProvider, operation));
+        }
+    }
+
+
+    public static boolean isRequestCompressionService(Model model, ServiceShape service) {
+        return TopDownIndex.of(model)
+                .getContainedOperations(service).stream()
+                .anyMatch(it -> it.hasTrait(RequestCompressionTrait.class));
+    }
+
+    private GoWriter.Writable writeMiddlewareHelper(SymbolProvider symbolProvider, OperationShape operation) {
+        var stackSymbol = SymbolUtils
+                .createPointableSymbolBuilder("Stack", SmithyGoDependency.SMITHY_MIDDLEWARE)
```

Comment:
`SmithyGoTypes` was merged with the SRA auth refactor - you can convert wrote non-generated imports like this to `SmithyGoTypes.Middleware.Stack` etc.

(add static symbols to `SmithyGoTypes` as needed)